### DISCUSSION TITLE
A typical use case currently not supported

### DISCUSSION THREAD

**Discussion by EricHerlitz** [postmanlabs/postman-flows#24] [Ideas]
This was written using Postman 9.0.5.

In any modern application we would rely on some sort of authentication token identifying the request and to my surprise variables seem to be dropped after the next block. This may be by design but it would make the entire thing less useful for most OpenID implementations or any SOA/MicroService app for that matter.

The illustrated example shows
1. Login, this will fetch a variable called authenticationToken
2. Send a request to fetch additional data from a login status endpoint using the authenticationToken
3. Create some variables to store the customer objects returned (the first variable step can be skipped), foreach to start a request iteration on each fetched customer object and a try to store the customerId as a variable, this works but only for one step
4. Get some subscriptionTypes connected to the customer, this is where we start to get into trouble, we have a customerId but the authenticationToken is empty. If I set it manually in the environment collection it will work, but that's sort of a no go.
5. Get more subscription information, this doesn't work at all since both customerId and authenticationToken is null.

![image](https://user-images.githubusercontent.com/1911851/137035003-4abfa945-e601-4762-bd3f-aba8adad460b.png)

The log will show something like this. Note that I've set the authenticationToken manually to have this running. I can also confirm the double entries as jucamo8311 found
![image](https://user-images.githubusercontent.com/1911851/137036326-54faeffe-0ad6-4df8-989a-db66a3f40662.png)

Also the subscriptionTypes api looks like this, and it does work if I enter the variables manually
![image](https://user-images.githubusercontent.com/1911851/137036636-9debe08b-1d23-4a34-a222-cbeafc48fa0a.png)

A workaround could possibly have been if the scripts worked, but either pre-request or test scripts seem to run at all when using Flows. 

Flows seem to have a great potential but until variable scopes like mentioned above the use cases is pretty narrow, at least for now.

**Comment by oehm-smith** [comment]
I think you meant at the very end ", but **neither** pre-request or test scripts seem to run at all when using Flows."

What happens if you 'prime' the customerId before setting it - perhaps defining ALL the variables to be used all at one time will make them global across the Flow. So in the first SetVariables block.

**Comment by samay-v** [comment]
Hi @EricHerlitz
As you said, this is by design. Blocks drop the data they receive unless explicitly said otherwise, like in case of a "Validate" or "Delay" block (AKA pass-through blocks).

That said, we understand the reusability of variables becomes an issue for use-cases like yours but there is a way to achieve it right now.

You can pipe the output from the first "Create Variable" block to multiple blocks (Shown in red). If you use the "Merge" block to merge the variables created from the 2 different "Create Variable" blocks (Pointed in blue) the requests will resolve both variables.

![Screenshot 2021-10-13 at 9 14 57 AM](https://user-images.githubusercontent.com/28357106/137066717-f8632d44-cdc8-4754-add7-1983da6bbb17.png)

This might become cumbersome for really big flows and are looking to see if a permanent store of data makes sense in Flows.

**Reply by oehm-smith** [reply]
Is there any ETA on the 'merge' and other new functionality being released in the desktop app?  I need this for the reasons discussed.  I can't use the WebApp as our APIs are not public.

I think the solution should be - set a scope on the variable.  I don't see the need to be able to set variables that affect running requests outside a flow (but I might be wrong) - just an 'immediate' (the current behaviour) and (global) 'flow' scope.

**Reply by EricHerlitz** [reply]
seems to be a more powerful version of the combine-block

**Reply by samay-v** [reply]
@oehm-smith I'll get back to you about the ETA for the desktop app. But just letting you know, you can use the [desktop agent](https://blog.postman.com/introducing-the-postman-agent-send-api-requests-from-your-browser-without-limits/) to use the webapp with local APIs

The current scope of variables that flows use are data variables, that do not affect the requests outside the flow.

**Reply by oehm-smith** [reply]
: (

I've never seen this MacOS message when trying to "open anyway".  I'm reluctant to proceed.

![image](https://user-images.githubusercontent.com/2556595/137828377-e4cba560-685f-4714-b7d9-9c7be4d2d2d5.png)

**Comment by Daveschofield** [comment]
I am having a similar issue where the requests in my flow rely on each other , data is generated dynamically within the requests. The data generated has to be unique. When starting the flow again i would expect the dynamic data to change as it  would the collection. I like the idea of flows as a tester, It  can relate the flow to the Business Analysts diagram for processing. Is there or will there be away to achieve this.

**Reply by anudeepmedicharla** [reply]
@Daveschofield Please tell me if I'm right/wrong. Do you want the ability to generate dynamic data within flows? A faker library perhaps?

**Reply by Daveschofield** [reply]
That is correct, currently i have a collection that has a series of requests. This drives a web front end that members of the public can sign up for and apply for various things 

I create a customer using the faker library and use a pre-request script to generate a random email address. We also have endpoints that will appear as drop down boxes in the web front end. I am calling the Endpoint and then grabbing a random value from the returned list and storing it as a variable to use further down the application process. 

![image](https://user-images.githubusercontent.com/28944622/138049733-fb06979f-8991-47d5-ba5a-e8cecb761e8c.png)


One of the endpoints returns a list of ethnicities, i get a random one from the returned json using the below 

//How mand ethinicities was returned 
   var response = JSON.parse(responseBody); 
   var count =  (Object.keys(response.data.values).length);
   console.log(count);

//get a random ethinicity in Array
   var ethinicity = randInt(count, 1);
   console.log(ethinicity);

//Retrieve the random ethnicity
let jsonData = pm.response.json();
console.log(jsonData.data.values[ethinicity]);
pm.environment.set("ethinicity", jsonData.data.values[ethinicity]);


function randInt(maximum, minimum) {
   return Math.round((Math.random() * Math.abs(maximum - minimum)) + minimum);
}

Hope this helps

**Reply by anudeepmedicharla** [reply]
I'm guessing we need 2 things:

1. Faker library as blocks - for now, you can use the requests stored within a collection but faker within flows sounds like an idea.
2. Be able to pick random items from a list instead of choosing a constant value.

**Reply by Daveschofield** [reply]
It would be nice for that feature also with the option of the start being able to do a completely fresh request

**Comment by oehm-smith** [comment]
Hi PostManLabs, sorry to ask again but this proved so promising I'd started essential testing but have stalled half way through due to this issue.  The WebApp version seems to have the functionality that fixes this issue - any idea of when that will make it in to the desktop app?

**Reply by samay-v** [reply]
https://github.com/postmanlabs/postman-flows/discussions/24#discussioncomment-1497220

**Comment by HenriAugusto** [comment]
I've came here looking for this exact use-case.

ps: for a better title it could be "A typical use case currently not supported (variables persistency)" or something like that?

**Reply by EricHerlitz** [reply]
Good idea, done.

**Comment by BrianJonesNMI** [comment]
When will merge be in the latest desktop build? I need to access non-public interfaces and the desktop agent is not working for me.

Thanks

**Reply by oehm-smith** [reply]
I updated to the latest desktop version and Yipee, it seems to have the 'merge' function!  Unfortunately the Flows I wrote are not loading!  They load fine in the webapp version.

...

https://support.postman.com/hc/en-us/articles/4410785125271#h_01FJBXQM8B6RXQB09WVECY2V85 seemed to solve it (How to remove local data)

**Reply by oehm-smith** [reply]
Nope I'm still having problems.  After deleting local data I restart PM and I can see the flows.  Now I quit and restart and I can no longer see them.  Interestingly it is only the flow that has the new 'merge' request.  All the others that didn't have that can be displayed just fine.

I will lodge a support ticket for this.
